llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td

```diff
@@ -435,6 +433,98 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
+  DeclareOpInterfaceMethods<ReifyRankedShapedTypeOpInterface>,
+  DeclareOpInterfaceMethods<LinalgExtInterface>,
+  DeclareOpInterfaceMethods<TilingInterface,
+    ["getIterationDomain",
+     "getLoopIteratorTypes",
+     "generateScalarImplementation"]>,
+  DeclareOpInterfaceMethods<MemoryEffectsOpInterface>
+]>{
+  let summary = "pack operation"; 
+  let description = [{
+    The pack operation takes a shaped type as input, tiles, and transposes it
+    based on `dims_pos` and `inner_tiles`. The former carries the indices of the
+    dimension to tile, while the latter the tiling factor. A tile factor can
+    either be a constant or an index type. We handle only full tiles now; it is UB
+    if the tile does not perfectly divide the dimension. We will relax this
+    constraint in the future when we add pad semantics to the operation.
+  }];
+
+  let arguments = (ins Variadic<AnyShaped>:$inputs,
+    Variadic<AnyShaped>:$outputs,
+    DefaultValuedAttr<I64ArrayAttr, "{}">:$dims_pos,
+    Variadic<Index>:$inner_tiles,
+    I64ArrayAttr:$static_inner_tiles);
+
+  let results = (outs Variadic<AnyRankedTensor>:$results);
+  let assemblyFormat = [{
+    attr-dict
+    $inputs 
+    `dims_pos` `=` $dims_pos
+    `inner_tiles` `=`
+    custom<DynamicIndexList>($inner_tiles, $static_inner_tiles,
+                             "ShapedType::kDynamicSize")
+    `into` $outputs `:` `(` type($inputs) type($outputs) `)`
+     (`->` type($results)^)?
+  }];
+
+  let extraClassDeclaration = extraLinalgExtOpClassDeclaration # [{
+
+    // Return the output operand.
+    Value getOutput() {
+      return getOutputOperand(0)->get();
+    }
+
+    // Return the input operand.
+    Value getInput() {
+      return getInputOperand(0)->get();
+    }
+
+    // Return the output rank.
+    int64_t getOutputRank() {
+      return  getOutput().getType().cast<ShapedType>().getRank();
+    }
+
+    // Return the output type.
+    ShapedType getOutputType() {
+      return getOutput().getType().cast<ShapedType>();
+    }
+
+    // Return the input type.
+    ShapedType getInputType() {
+      return getInput().getType().cast<ShapedType>();
+    }
+
+    // Return the output shape.
+    ArrayRef<int64_t> getOutputShape() {
+      return getOutputType().cast<ShapedType>().getShape();
+    }
+
+    // Return the input shape.
+    ArrayRef<int64_t> getInputShape() {
+      return getInputType().cast<ShapedType>().getShape();
+    }
+
+    // Return the rank of the input operand.
+    int64_t getInputRank() {
+      return getInput().getType().cast<ShapedType>().getRank();
+    }
+
+    // Return the tile sizes.
+    SmallVector<OpFoldResult> getMixedTiles();
+    SmallVector<int64_t> getStaticTiles();
+
+    // Infer the output type.
+    ShapedType inferResultType();
+
+    // Return a mapping from positions `dims_pos` to their tile factors.
+    DenseMap<int64_t, OpFoldResult> getDimAndTileMapping();
+    DenseMap<int64_t, int64_t> getDimAndStaticTileMapping();
```

Comment:
I am not sure we need the `getDimAndStaticTileMapping`. The `getDimAndTileMapping` has everything we need I think.

Reply:
removed.